vm/actor/src/builtin/miner/expiration_queue.rs

```diff
@@ -0,0 +1,785 @@
+        self.amt
+            .get(key as u64)
+            .map_err(|e| format!("failed to lookup queue epoch {}: {:?}", key, e))?
+            .ok_or_else(|| format!("missing expected expiration set at epoch {}", key))
+    }
+
+    fn must_update(
+        &mut self,
+        epoch: ChainEpoch,
+        expiration_set: ExpirationSet,
+    ) -> Result<(), String> {
+        self.amt
+            .set(epoch as u64, expiration_set)
+            .map_err(|e| format!("failed to set queue epoch {}: {:?}", epoch, e))
+    }
+
+    /// Since this might delete the node, it's not safe for use inside an iteration.
+    fn must_update_or_delete(
+        &mut self,
+        epoch: ChainEpoch,
+        expiration_set: ExpirationSet,
+    ) -> Result<(), String> {
+        if expiration_set.is_empty() {
+            self.amt
+                .delete(epoch as u64)
+                .map_err(|e| format!("failed to delete queue epoch {}: {:?}", epoch, e))?;
+        } else {
+            self.amt
+                .set(epoch as u64, expiration_set)
+                .map_err(|e| format!("failed to set queue epoch {}: {:?}", epoch, e))?;
+        }
+
+        Ok(())
+    }
+}
+
+struct SectorEpochSet {
+    epoch: ChainEpoch,
+    sectors: Vec<u64>,
+    power: PowerPair,
+    pledge: TokenAmount,
+}
+
+/// Takes a slice of sector infos and returns sector info sets grouped and
+/// sorted by expiration epoch, quantized.
+///
+/// Note: While the result is sorted by epoch, the order of per-epoch sectors is maintained.
+fn group_sectors_by_expiration<'a>(
+    sector_size: SectorSize,
+    sectors: impl IntoIterator<Item = &'a SectorOnChainInfo>,
+    quant: QuantSpec,
+) -> Vec<SectorEpochSet> {
+    let mut sectors_by_expiration = HashMap::<ChainEpoch, Vec<&SectorOnChainInfo>>::new();
+
+    for sector in sectors {
+        let q_expiration = quant.quantize_up(sector.expiration);
+        sectors_by_expiration
+            .entry(q_expiration)
+            .or_default()
+            .push(sector);
+    }
+
+    // This map iteration is non-deterministic but safe because we sort by epoch below.
+    let mut sector_epoch_sets: Vec<_> = sectors_by_expiration
+        .into_iter()
+        .map(|(expiration, epoch_sectors)| {
+            let mut sector_numbers = Vec::<u64>::with_capacity(epoch_sectors.len());
+            let mut total_power = PowerPair::zero();
+            let mut total_pledge = TokenAmount::zero();
+
+            for sector in epoch_sectors {
+                sector_numbers.push(sector.sector_number);
+                total_power += &power_for_sector(sector_size, &sector);
+                total_pledge += &sector.initial_pledge;
+            }
+
+            SectorEpochSet {
+                epoch: expiration,
+                sectors: sector_numbers,
+                power: total_power,
+                pledge: total_pledge,
+            }
+        })
+        .collect();
+
+    sector_epoch_sets.sort_by_key(|epoch_set| epoch_set.epoch);
```

Comment:
Isn't it possible for two sectors to have the same expiry epoch? In which case we may have a discrepancy with the hashmap ordering differences between rust/go. Thoughts?

Reply:
I'm not sure I understand but I think you misread this function, the purpose of `sectors_by_expiration` is to group sectors together with the same expiry epoch. So that guarantees that the `SectorEpochSet`s in `sector_epoch_sets` have distinct epochs.